### PR TITLE
Implementing additional SHA256-based hasher

### PR DESCRIPTION
**Pull request by storojs72**
This is a very first PR from my side. It has following goals:
- to get taste of actual communication/collaboration with the team;
- get deeper understanding of the merkletree codebase including small details;
- start unit-testing of SHA256 based Merkle Tree that is used in Filecoin 32Gb sectors (in production).

Here I added new hasher based on SHA-256 additionally to the existing XOR128. I re-used unit tests (by making them generic) specified mostly in test_xor128.rs source file which seems to be the main point of testing. For simplicity I used only trees with `VecStore` storage layer and extending tests to other storages is left for future work.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -41,7 +41,7 @@ byteorder = "1.3.1"
 env_logger = "0.7.1"
 
 [features]
-default = []
+default = ["rust-crypto"]
```
Comment (vmx) [inline Cargo.toml:44]: Is this a left-over from debugging, or should `rust-crypto` really be a default reagure?

File: src/test_common.rs [hunk 2]
```diff
@@ -53,22 +59,6 @@ impl Algorithm<Item> for XOR128 {
     }
 }
 
-impl fmt::UpperHex for XOR128 {
```
Comment (vmx) [inline src/test_common.rs:56]: Why did you remove that formatting implementation?

File: src/test_common.rs [hunk 3]
```diff
@@ -86,9 +76,64 @@ impl Element for [u8; 16] {
...
+    pub fn new() -> Sha256Hasher {
+        Sha256Hasher {
+            engine: Sha256::new(),
+        }
+    }
+}
+
+impl Debug for Sha256Hasher {
+    fn fmt(&self, _f: &mut Formatter<'_>) -> fmt::Result {
+        todo!()
```
Comment (vmx) [inline src/test_common.rs:97]: This will panic if you want to print a `Sha256Hasher`. You could either remove the implementation completely, or (even better) implement some debug output.

File: src/test_common.rs [hunk 4]
```diff
@@ -86,9 +76,64 @@ impl Element for [u8; 16] {
...
+impl Default for Sha256Hasher {
+    fn default() -> Self {
+        Sha256Hasher::new()
+    }
+}
+
+impl Hasher for Sha256Hasher {
+    // FIXME: contract is broken by design
+    fn finish(&self) -> u64 {
+        unimplemented!()
```
Comment (vmx) [inline src/test_common.rs:110]: FYI: `unimplemented()` also supports adding a message, so perhaps putting the information from the comment into the `unimplemented!()` could make sense.

File: src/test_common.rs [hunk 5]
```diff
@@ -86,9 +76,64 @@ impl Element for [u8; 16] {
...
+    }
+
+    fn write(&mut self, bytes: &[u8]) {
+        self.engine.input(bytes)
+    }
+}
+
+impl Algorithm<Item> for Sha256Hasher {
+    fn hash(&mut self) -> Item {
+        let mut result: Item = Item::default();
```
Comment (vmx) [inline src/test_common.rs:120]: FYI (no need to change). You can also use `let mut result = Item::default()`, which doesn't specify the type. Or `let mut result: Item = Default::default()`, which doesn't specify the type when calling `default()`.

File: src/test_xor128.rs [hunk 6]
```diff
@@ -682,56 +782,46 @@ fn test_compound_levelcache_octrees_trees_from_store_configs() {
     test_compound_levelcache_tree_from_store_configs::<U8, U7>(4096);
 }
 
-#[test]
```
Comment (vmx) [inline src/test_xor128.rs:685]: What happened to this test?

**Review by vmx — COMMENTED** [review]
Overall it looks good, I just have questions/commented on minor things.

File: Cargo.toml (hunk 1 above)
Comment (storojs72) [reply Cargo.toml:44]: This is because we need to take SHA-256 from some "crypto-provider" crate in order to implement Sha256Hasher. Previously we didn't need it in default feature since the only hasher was XOR128 and it didn't use any crypto under the hood. I decided to take `rust-crypto`, because it looks "more transparent" then `ring` one, which actually involves `boringssl` C library.

File: src/test_common.rs (hunk 2 above)
Comment (storojs72) [reply src/test_common.rs:56]: We used it only for `test_hasher_light` where we hashed strings. Now this test is rewritten and it uses byte vectors which seems to be standard practice in crypto (passing strings to cryptographic algorithms always requires dealing with encoding which is actually additional headache). Apart from that I didn't actually see where this logic is needed

File: src/test_common.rs (hunk 3 above)
Comment (storojs72) [reply src/test_common.rs:97]: Right. I can't remove it since linter swears on it (`missing_debug_implementations` is enabled). So, I will add this implementation

File: src/test_common.rs (hunk 5 above)
Comment (storojs72) [reply src/test_common.rs:120]: Right, thanks! I just prefer to explicitly state all types always

File: src/test_xor128.rs (hunk 6 above)
Comment (storojs72) [reply src/test_xor128.rs:685]: Thanks for paying attention on this. It seems to me that this test is actually duplicated by `test_compound_quad_trees_from_slices`. We discussed with @cryptonemo, that it makes sense to use generic test like this:

```
fn test_tree() {
   fn run_test() {
      ...
   }
   run_test::<_, _, U8>
   run_test::<_, _, U4>
}
``` 

instead of two tests for quad and oct trees with actually similar logic:
```
fn test_oct_tree() {}
fn test_quad_tree() {}
```

In this PR I didn't introduced this change - only manually analysed what tests are redundant. This will be introduced in future PRs.

File: src/test_common.rs (hunk 4 above)
Comment (storojs72) [reply src/test_common.rs]: Ok! Added in https://github.com/filecoin-project/merkletree/pull/97/commits/45019d5d5727e5211e2c8c979958eb1a52f6f61e

File: src/test_common.rs (hunk 3 above)
Comment (storojs72) [reply src/test_common.rs]: Added in https://github.com/filecoin-project/merkletree/pull/97/commits/45019d5d5727e5211e2c8c979958eb1a52f6f61e

File: Cargo.toml (hunk 1 above)
Comment (vmx) [reply Cargo.toml:44]: `rust-crypto` seems to be used only by tests and benchmarks. So I propose making it a non-optional dev dependency instead. This way it won't be part of the dependency tree if you have `merkletree` as a dependency.

File: src/test_common.rs (hunk 2 above)
Comment (vmx) [reply src/test_common.rs:56]: As `XOR128` is only used for testing and not part of the public interface it's OK to remove that debug formatting.

**Review by vmx — CHANGES_REQUESTED** [review]
There's only one thing left I'd like to see a change on, the rest looks good!

File: Cargo.toml (hunk 1 above)
Comment (storojs72) [reply Cargo.toml:44]: It seems that we can't do that without significant reorganisation of the codebase actually. If we make `rust-crypto` a dev-dependency, than we will need to move (and start thinking about them as integration tests, which they are by their nature) tests to /tests directory from the following list of files:
- test_cmh.rs
- test_common.rs
- test_item.rs
- test_sip.rs
- test_xor128.rs

which "historically" was considered as unit-tests and was actually a part of merkletree source code (located in `/src` directory under `#cfg(test)` macro).

But in general I agree that this change should be introduced because source code of this repository is not an appropriate place for actual hasher's implementations. E.g. in merkletree repository we provide hasher's abstractions and interfaces but implementing hasher is considered to be a task for external caller (like [rust-file-proofs](https://github.com/filecoin-project/rust-fil-proofs/tree/75d6fd8b96cfa85821b56a8bf59b88882d446015/filecoin-hashers/src) caller, where we have Blake2s, Poseidon and Sha256 hashers). 

This seems to be a rather big change which I would make as a next separate incremental PR. WDYT?

Comment (cryptonemo) [reply Cargo.toml:44]: Separate PR sounds fine, thanks

Comment (vmx) [reply Cargo.toml:44]: I agree that a separate PR for re-organizing the tests makes sense and also that this re-organizing should happen. Though I don't think it's needed. As long as the tests are behind `#![cfg(test)]`, it should work having `rust-crypto` as dev dependency.

**Comment by storojs72** [comment]
@vmx, thank you for reviewing!

Regarding conversation about `rust-crypto` dependency, I agree that by default we should not require any crypto-providers as library dependencies, because it is a task for high-level caller (like `rust-fit-proofs`) to implement particular hasher. At the same time, I think that this is also true for `ring` dependency. It seems that introducing features to this library is actually legacy from original [repository](https://github.com/sitano/merkle_light) and it has no relevance to how this library is utilised by Filecoin. I would propose introducing reorganisation from https://github.com/filecoin-project/merkletree/pull/97#discussion_r832086949 and removing legacy features unless you or other engineers are aware whether it is needed somewhere at higher level.

**Review by vmx — APPROVED** [review]
As I agree that the re-organization should be a separate PR, I approve this one.

File: src/test_common.rs [hunk 7]
```diff
@@ -1,3 +1,4 @@
+#[cfg(test)]
```
Comment (vmx) [inline src/test_common.rs:1]: You can also use `#![cfg(test)]` (note the esclamation mark after the hash). This means "put the whole file under the `cfg(test)` flag. Without the exclamation mark it only applies to the next line. This way you also don't need the `#[cfg(test)]` in `src/lib.rs`.

File: Cargo.toml (hunk 1 above)
Comment (vmx) [reply Cargo.toml]: Yes exactly! See my other comment, that you need to use `#![cfg(test)]` so that it is applied to the whole file.

File: src/test_common.rs (hunk 7 above)
Comment (storojs72) [reply src/test_common.rs:1]: Right. Thank you! And it seems to be fair for other test mods, so we can remove other `#[cfg(test)]` directives in `lib.rs` having those directive in appropriate source files

**Review by vmx — APPROVED** [review]
The CI is actually green: https://app.circleci.com/pipelines/github/filecoin-project/merkletree/205/workflows/5270c306-77c5-46ce-97dc-8a52976f09c4. It seems there was an issue with GitHub when it ran.
